cou/steps/plan.py

```diff
@@ -34,22 +34,264 @@
 from cou.apps.channel_based import ChannelBasedApplication  # noqa: F401
 from cou.apps.core import Keystone, Octavia  # noqa: F401
 from cou.apps.subordinate import SubordinateApplication, SubordinateBase  # noqa: F401
-from cou.commands import CLIargs
+from cou.commands import CONTROL_PLANE, DATA_PLANE, HYPERVISORS, CLIargs
 from cou.exceptions import (
+    DataPlaneCannotUpgrade,
+    DataPlaneMachineFilterError,
     HaltUpgradePlanGeneration,
     HighestReleaseAchieved,
     NoTargetError,
     OutOfSupportRange,
 )
-from cou.steps import PreUpgradeStep, UpgradePlan
+from cou.steps import PostUpgradeStep, PreUpgradeStep, UpgradePlan
 from cou.steps.analyze import Analysis
 from cou.steps.backup import backup
-from cou.utils.juju_utils import DEFAULT_TIMEOUT
+from cou.steps.hypervisor import HypervisorUpgradePlanner
+from cou.utils.app_utils import set_require_osd_release_option
+from cou.utils.juju_utils import DEFAULT_TIMEOUT, Machine, Unit
+from cou.utils.nova_compute import get_empty_hypervisors
 from cou.utils.openstack import LTS_TO_OS_RELEASE, OpenStackRelease
 
 logger = logging.getLogger(__name__)
 
 
+def pre_plan_sanity_checks(args: CLIargs, analysis_result: Analysis) -> None:
+    """Pre checks to generate the upgrade plan.
+
+    :param args: CLI arguments
+    :type args: CLIargs
+    :param analysis_result: Analysis result.
+    :type analysis_result: Analysis
+    """
+    verify_supported_series(analysis_result)
+    verify_highest_release_achieved(analysis_result)
+    verify_data_plane_ready_to_upgrade(args, analysis_result)
+    verify_hypervisors_cli_input(args, analysis_result)
```

Comment:
non-blocking: these, and many other functions, in this file can be marked as private since they are helper functions and only called from one place.